lib/paraactions/paraactions.ts

```diff
@@ -0,0 +1,790 @@
+      }
+
+      // Top-level whitespace separates actions
+      if (/\s/.test(ch) && depthParens === 0 && depthBraces === 0) {
+        break;
+      }
+    }
+
+    i++;
+  }
+
+  const raw = source.slice(startIndex, i).trim();
+  if (!raw) {
+    return { action: null, nextIndex: i };
+  }
+
+  const action = parseSingleAction(raw);
+  return { action, nextIndex: i };
+}
+
+// ─────────────────────────────────────────────────────────────────────────────
+// Parsing of Named / Chain actions (re-used inside blocks)
+// ─────────────────────────────────────────────────────────────────────────────
+
+/**
+ * Parse a single action string into either a NamedAction or a ChainAction.
+ */
+function parseSingleAction(input: string): ParaAction | null {
+  if (!input) return null;
+
+  const trimmed = input.trim();
+  if (!trimmed) return null;
+
+  // Chain call if there's a dot BEFORE the first '(' (e.g., api.getChart(...))
+  const firstParen = trimmed.indexOf('(');
+  const firstDot = trimmed.indexOf('.');
+  const isChain = firstDot !== -1 && (firstParen === -1 || firstDot < firstParen);
+
+  if (isChain) {
+    return parseChainAction(trimmed);
+  }
+
+  return parseNamedAction(trimmed);
+}
+
+function parseNamedAction(input: string): NamedAction | null {
+  const nameMatch = /^([a-zA-Z_][\w]*)/.exec(input);
+  if (!nameMatch) return null;
+
+  const name = nameMatch[1];
+  const rest = input.slice(name.length).trim();
+
+  let args: ParaActionArg[] = [];
+  if (rest.startsWith('(') && rest.endsWith(')')) {
+    const inner = rest.slice(1, -1);
+    args = parseArgs(inner);
+  }
+
+  return {
+    kind: 'namedAction',
+    name,
+    args,
+  };
+}
+
+function parseChainAction(input: string): ChainAction | null {
+  // Split by '.' but keep parentheses content together.
+  const segments: ChainSegment[] = [];
+  let current = '';
+  let depth = 0;
+  let inSingle = false;
+  let inDouble = false;
+
+  const flush = () => {
+    const trimmed = current.trim();
+    if (trimmed) {
+      const seg = parseChainSegment(trimmed);
+      if (seg) segments.push(seg);
+    }
+    current = '';
+  };
+
+  for (let i = 0; i < input.length; i++) {
+    const ch = input[i];
+
+    if (ch === '\'' && !inDouble) {
+      inSingle = !inSingle;
+      current += ch;
+      continue;
+    }
+    if (ch === '"' && !inSingle) {
+      inDouble = !inDouble;
+      current += ch;
+      continue;
+    }
+
+    if (!inSingle && !inDouble) {
+      if (ch === '(') {
+        depth++;
+        current += ch;
+        continue;
+      }
+      if (ch === ')') {
+        depth = Math.max(0, depth - 1);
+        current += ch;
+        continue;
+      }
+      if (ch === '.' && depth === 0) {
+        flush();
+        continue;
+      }
+    }
+
+    current += ch;
+  }
+
+  flush();
+
+  if (segments.length === 0) {
+    return null;
+  }
+
+  return {
+    kind: 'chain',
+    segments,
+  };
+}
+
+function parseChainSegment(segment: string): ChainSegment | null {
```

Comment:
If we merge `NamedAction` and `ChainSegment` then this function won't be needed.